System/Random.hs

```diff
@@ -29,104 +29,38 @@
+-- The original library was and still is split into two layers:
 --
--- * A core /random number generator/ provides a supply of bits.
---   The class 'RandomGen' provides a common interface to such generators.
+-- * A core /random number generator/ provides a supply of bits.  The
+--   class 'RandomGen' provides a common interface to such generators.
 --   The library provides one instance of 'RandomGen', the abstract
---   data type 'StdGen'.  Programmers may, of course, supply their own
+--   data type 'StdGen'.  This implementation uses the SplitMix
+--   algorithm [1]. Programmers may, of course, supply their own
 --   instances of 'RandomGen'.
 --
 -- * The class 'Random' provides a way to extract values of a particular
 --   type from a random number generator.  For example, the 'Float'
 --   instance of 'Random' allows one to generate random values of type
 --   'Float'.
 --
--- This implementation uses the SplitMix algorithm [1].
+-- However, this, e.g. @random :: RandomGen g => g -> (a, g)@,
+-- interface means that any instance for @a@ must be able to use any
+-- generator @g@ forcing all such instances to use infinite precision
+-- 'Integer' with a consequent performance impact.
 --
--- [/Example for RNG Implementors:/]
---
--- Suppose you want to use a [permuted congruential
--- generator](https://en.wikipedia.org/wiki/Permuted_congruential_generator)
--- as the source of entropy (FIXME: is that the correct
--- terminology). You can make it an instance of `RandomGen`:
---
--- >>> data PCGen = PCGen !Word64 !Word64
---
--- >>> :{
--- let stepGen :: PCGen -> (Word32, PCGen)
---     stepGen (PCGen state inc) = let
---       newState = state * 6364136223846793005 + (inc .|. 1)
---       xorShifted = fromIntegral (((state `shiftR` 18) `xor` state) `shiftR` 27) :: Word32
---       rot = fromIntegral (state `shiftR` 59) :: Word32
---       out = (xorShifted `shiftR` (fromIntegral rot)) .|. (xorShifted `shiftL` fromIntegral ((-rot) .&. 31))
---       in (out, PCGen newState inc)
--- :}
---
--- >>> fst $ stepGen $ snd $ stepGen (PCGen 17 29)
--- 3288430965
---
--- >>> :{
--- instance RandomGen PCGen where
---   next g = (fromIntegral y, h)
---     where
---       (y, h) = stepGen g
---   split _ = error "This PRNG is not splittable"
--- :}
+-- Thus, implementors are strongly encouraged to provide efficient
+-- versions of at least 'genWord32' and 'genWord64'.
 --
--- Importantly, this implementation will not be as efficient as it
--- could be because the random values are converted to 'Integer' and
--- then to desired type.
+-- FIXME: I want to write that if the implementor just supplies 'next'
+-- then defaults are provided but these will go via an function
+-- (@randomIvalInteger@ which will not be named) which will use the
+-- available entropy to produce a 32 bit random value but will use
+-- 'Integer' to do this resulting in inefficient generation. BUT are
+-- we going to keep @randomIvalInteger@????
```

Comment:
I think that for the purpose of this PR, we can assume that `randomIvalInteger` stays. The alternative would be something like https://github.com/idontgetoutmuch/random/pull/43, but we can change the docs there if we need to.